spec/mail/mailbox.mdown

```diff
@@ -34,6 +34,27 @@ A **Mailbox** object has the following properties:
   0 <= sortOrder < 2^31.
 
     A mailbox with a lower order should be displayed before a mailbox with a higher order (that has the same parent) in any mailbox listing in the client's UI. Mailboxes with equal order SHOULD be sorted in alphabetical order by name. The sorting SHOULD take into account locale-specific character order convention.
+- **sharedWith**: `Rights[String]`
```

Comment:
I'm not convinced this should be part of the core spec; I think there's enough extra complexity and domain-specific issues involved to make setting up sharing an extension. Some of the many issues with this:

* Who may I share a Mailbox with? What happens if I try to specify someone else? What happens when the set of users I am permitted to share with changes?
* What happens if that user is renamed? (Ids are preferred to "names" for references in JMAP for this reason.)
* If I share mailbox X but not mailbox Y, and there exists a message Z in both of these mailboxes, what does the sharee see as the *mailboxIds* property when they fetch that message? What happens if I then share mailbox Y with them; is there a change to the state string for Messages?
* If a message is in mailbox X and Y, and X is shared to me with the "w" ACL, but Y is shared without the "w" ACL, may I modify the keywords of a message that exists in both mailboxes?
* Why do the set of rights not match the mayXXX properties of the Mailbox object?

Reply:
I Neil,

I will  answer a few of your remarks on it...

```
If I share mailbox X but not mailbox Y, and there exists a message Z in both of these mailboxes, what does the sharee see as the mailboxIds property when they fetch that message? What happens if I then share mailbox Y with them; is there a change to the state string for Messages?
```

I don't see what prevent me to have such kind of "similar problem" with the current spec. may\* properties, and partial acount sharing are enough to create these kind of situations.

The answer is:
 - mailboxIds one don't have "mayReadItems" on it are filtered out of mailboxId property

Similar concerns are also observable with flags. Which is your next concern: 

```
If a message is in mailbox X and Y, and X is shared to me with the "w" ACL, but Y is shared without the "w" ACL, may I modify the keywords of a message that exists in both mailboxes?
```

Again, these concerns are not answered by "may\*" properties.

I believe that by giving delegation to write flags to a mailbox, we give the writes to update the flags of the message objects it references. Thus it seems logical to me to update flags even in mailboxes the user can not if at least one mailbox allows it.

Note that, as a user, the way to solve this is to use **importMessages** when appending email on a shared mailbox.

So IMO **These mailboxId and keywords concerns are orthogonal to right setting logic** as they are linked with the notion of right itself, already introduced in the spec. I don't see why it should abort this PR.

--------------------

```
Why do the set of rights not match the mayXXX properties of the Mailbox object?
```

I think syntactical changes are not the hardest one here, and can very well be adapted

----------------

Concerning the designation of a user, I believe you then think about using **accountId** but that's a "tricky question" as you might end up doing some disclosure by validating / invaidating the existance of specific user. 

By choosing email address, we allow one to share it to the other user if they know the email address, thus if they already had some kind of exchange.  However the "rename support" is IMO a real concern, at least to include this in the JMAP core specs.

What are your thoughts on it?

 - Allowing mailAddress -> accountId search would validate the existance of the mailAddress user. What of course is a security cncern.
 - Without this, we can not allow arbitrary sharing...

(and we can not base the accountId on the mailAddress (IE: SHA-512) as it would defeat the address change).

Some thoughts on it ?

Reply:
Hi,

I will try to answer more specifically to some of your questions.

> Who may I share a Mailbox with? What happens if I try to specify someone else? What happens when the set of users I am permitted to share with changes?

Choosing to share with a mail address solves many of these concerns, but I agree we must handle some error cases for this, without discarding too much information.

> What happens if that user is renamed? (Ids are preferred to "names" for references in JMAP for this reason.)

This could be handled at the implementation level: user shares his mailbox with an email address, but implementation map this email to an accountId and store the share related to the accountId, not the email. So when the user is renamed, the sharing remains.

> If I share mailbox X but not mailbox Y, and there exists a message Z in both of these mailboxes, what does the sharee see as the mailboxIds property when they fetch that message? What happens if I then share mailbox Y with them; is there a change to the state string for Messages?

I think this has been answered by Benoit. About the state string I'm not sure, as we don't implement it for now, but at first sight I would say "yes".

> Why do the set of rights not match the mayXXX properties of the Mailbox object?

You are right on this, but mayXXX properties are a little poor to express some concerns. For example the lookup right only is designed by all the mayXXX properties being false. Maybe changing some mayXXX properties and mapping sharedWith to them could be the way to go.

Regards,
Raphaël Ouazana.

Reply:
I tried to do the mapping from rights to mayXXX property, here is my proposal:
l => mayRead (new)
r => mayReadItems
w => mayModifyItems (new)
i => mayAddItems
k => mayCreateChild
x => mayDelete
t => mayMoveToTrashItems (new)
e => mayRemoveItems
a => mayAdminister (new)
As you can see it introduces 4 new rights.
Furthermore, mayRename should be IMHO expressed as mayCreateChild + mayDelete (as in IMAP specification) and so could be removed, or just be kept as a read-only property.
In the other way mayRead could be removed as a read-only property, because by definition it is always true. But it can be interesting to keep it for coherence.

If you agree with this, I will update this PR with this change, including adding new mayXXX properties.

Reply:
FYI, there's some discussion of this in #126